cmd/promlts/query.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"net"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/promlts/pkg/query"
+	"github.com/oklog/oklog/pkg/group"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/route"
+	"github.com/prometheus/prometheus/promql"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+// registerQuery registers a query command.
+func registerQuery(app *kingpin.Application, name string) runFunc {
+	cmd := app.Command(name, "query node exposing PromQL enabled Query API with data retrieved from multiple store nodes")
+
+	apiAddr := cmd.Flag("api-address", "listen address for the query API").
+		Default(":19099").String()
+
+	storeAddresses := cmd.Flag("store.addresses", "comma delimited listen addresses of store APIs").
+		Default("localhost:19090").Strings()
+
+	queryTimeout := cmd.Flag("query.timeout", "maximum time to process query by query node").
+		Default("2m").Duration()
+
+	maxConcurrentQueries := cmd.Flag("query.max-concurrent", "maximum number of queries processed concurrently by query node").
+		Default("20").Int()
+
+	return func(logger log.Logger, metrics prometheus.Registerer) error {
+		return runQuery(logger, metrics, *apiAddr, query.Config{
+			StoreAddresses:       *storeAddresses,
+			QueryTimeout:         *queryTimeout,
+			MaxConcurrentQueries: *maxConcurrentQueries,
+		})
+	}
+}
+
+// runQuery starts a server that exposes PromQL Query API. It is responsible for querying configured
+// store nodes, merging and duplicating the data to satisfy user query.
+func runQuery(
+	logger log.Logger,
+	reg prometheus.Registerer,
+	apiAddr string,
+	cfg query.Config,
+) error {
+	level.Info(logger).Log(
+		"msg", "I'm a query node",
+		"api-address", apiAddr,
+		"store.addresses", cfg.StoreAddresses,
+		"query.timeout", cfg.QueryTimeout,
+		"query.max-concurrent", cfg.MaxConcurrentQueries,
```

Comment:
We should probably just have a simple message for each command that it started. Probably we should move them right before `g.Run` as everything before is setup.

Reply:
Well, is it bad to have few more arguments with config values? I will move that after the `group` setup.